## (m) solution:

In [1]:
# Question (m)

# Apply PDE to price the given barrier call option

import numpy as np

# define parameters
x, K = 100., 110. # initial price, strike price
B, R = 80., 300. # barrier, big constant value
t, T = 0., 1. # initial time, maturity
Nx, Nt = 2200, 252 # number of price grid, time grid
r, sigma = .02, .2 # drift rate, volatility

# compute small price interval and time interval
dx = (R - B) / Nx
dt = (T - t) / Nt

# construct M given by question (l)
M = np.zeros((Nx + 1, Nx + 1))
for k in range(Nx + 1):
    if (k == 0 or k == Nx):
        M[k][k] = 1
    else:
        M[k][k] = 1 + r*dt + ((sigma**2 * (B + k*dx)**2 * dt) / (dx ** 2))
        M[k][k + 1] = -((r*(B + k*dx))/(2*dx) + (sigma**2 * (B + k*dx)**2)/(2 * dx**2))*dt
        M[k][k - 1] = ((r*(B + k*dx))/(2*dx) - (sigma**2 * (B + k*dx)**2)/(2 * dx**2))*dt

# construct V_{Nt}, vector for final conditions
Vcurrent = np.zeros(Nx + 1)
for k in range(Nx + 1):
    Vcurrent[k] = max((B + k*dx) - K, 0)

invM = np.linalg.inv(M)
# apply recursive relation to compute V0
for j in range(Nt - 1, -1, -1):
    Vnext = np.matmul(invM, Vcurrent)
    Vnext[0] = 0
    # boundary correction
    Vnext[-1] = R - np.exp(r*(-(T - j*dt)))*K
    Vcurrent = Vnext
    
# choose the value in V0 that represent the price of call 
# when initial price is 100, and display the result
print("The price of call solved by PDE with initial price $100 is: ", Vcurrent[200])

The price of call solved by PDE with initial price $100 is:  4.916733625244888


From the code and the result showed above, we know that the PDE solution of this call price is 4.916733625244888.